### ISSUE TITLE
freetype-2.11.0 and terminus

### ISSUE THREAD

**Issue by marrowsuck** [alacritty/alacritty#5353]
I just upgraded to freetype-2.11.0 and alacritty fails to start with the terminus font.
freetype-2.10.4 with terminus and alacritty works fine.

```
[2021-07-20 23:17:38.706285472] [DEBUG] [crossfont] Loaded Face Face { ft_face: Font Face: Regular, load_flags: DEFAULT | TARGET_NORMAL, render_mode: "Normal", lcd_filter: 1 }
[2021-07-20 23:17:38.708456515] [ERROR] [alacritty] Alacritty encountered an unrecoverable error:
                                                    
                                                        Cannot render glyph
```

**Comment by chrisduerr** [comment]
This looks like a freetype error. Have you tried any different programs using freetype for text rendering? It's possible there was just a change in supported formats in freetype.

**Comment by andyrtr** [comment]
That's the same behavior I see here testing Arch Linux packages with Terminus and Smoothansi.se and probably other otb fonts. Other terminals like urvxt or Xfce4-terminal keep working with new freetype release.
No difference here with alacritty 0.8.0 or a current git build.

**Comment by chrisduerr** [comment]
If a non-breaking upgrade of freetype breaks something, I don't see why there would be an issue with Alacritty. I'd have to look at what exactly the issue is, though the error definitely doesn't come from Alacritty.

**Comment by DarkDefender** [comment]
This seems to indeed be an issue with Freetype.

If I downgrade to freetype 2.10.4 it work.

Freetype 2.11.0 also breaks the `foot` terminal.
In there I prints where if fails: https://codeberg.org/dnkl/fcft/src/branch/master/fcft.c#L1268

It seems like freetype 2.11.0 at least reports that rendering fails. I also get error messages in Qt that rendering failed, but the font seems to be displayed correctly regardless which makes be a bit confused.

Perhaps it is a bug in freestyle that it reports that it failed to render with bitmap fonts even if it didn't fail?

**Comment by chrisduerr** [comment]
Closing the Alacritty ticket for now. There seems to be no indication that Alacritty should be doing something differently.

**Comment by DarkDefender** [comment]
I've created a bug report upstream: https://gitlab.freedesktop.org/freetype/freetype/-/issues/1076

**Comment by nmeum** [comment]
> There seems to be no indication that Alacritty should be doing something differently.

Foot added an explicit workaround for this API change in freetype-2.11.4 with foot-2.4.4. See https://codeberg.org/dnkl/fcft/commit/f645bc18ce6ce1c14aa1b448b095e0f5915ddd0b I personally think it wouldn't be unreasonable to add a similar change to Alacritty.

Consider the following: With freetype < 2.11.4 the `FT_Render_Glyph` does not return an error when it is used to render bitmap fonts. However, with freetype >= 2.11.4 an error is returned in this case. From the freetype perspective "rendering" Bitmap fonts with `FT_Render_Glyph` [is a no-op](https://gitlab.freedesktop.org/freetype/freetype/-/issues/1076#note_1000444) (it doesn't do anything). That's why the foot change linked above simply does not call `FT_Render_Glyph` for Bitmap fonts which fixes this issue. The same thing could easily be done in Alacritty as well and given that this operation seems to be a no-op for bitmap fonts it doesn't seem unreasonable to me to add a foot-like case-distinction to Alacritty as well.

**Comment by chrisduerr** [comment]
Thanks a lot for the info @nmeum. That's definitely a breaking change from freetype though which is somewhat surprising.

I'd have to test things myself, but I'd assume foot's approach should work just fine.

**Comment by WhyNotHugo** [comment]
Any idea on when this will be released?

I had pinned an older freetype as a workaround, but now other software has issues with this older freetype. :(

**Comment by kchibisov** [comment]
It's already included in `v0.9.0` release

**Comment by WhyNotHugo** [comment]
It is. My bad, sorry for the noise.
